obsidianirc.c

```diff
@@ -0,0 +1,640 @@
+        }
+    }
+    return NULL;
+}
+
+/**
+ * accreg_capability_parameter - Returns the parameter string for the registration capability.
+ */
+const char *accreg_capability_parameter(Client *client)
+{
+    return "before-connect,custom-account-name,email-required";
+}
+
+/**
+ * accreg_capability_visible - Determines if the registration capability is visible to the client.
+ */
+int accreg_capability_visible(Client *client)
+{
+    return 1;
+}
+
+/**
+ * open_database - Opens the SQLite database.
+ */
+int open_database(const char *filename)
+{
+    if (sqlite3_open(filename, &db) != SQLITE_OK)
+    {
+        return SQLITE_ERROR;
+    }
+    const char *sql =   "CREATE TABLE IF NOT EXISTS accounts ("
+                        "id INTEGER PRIMARY KEY AUTOINCREMENT, "
+                        "name TEXT, "
+                        "email TEXT, "
+                        "password TEXT, "
+                        "time_registered INTEGER, "
+                        "verified INTEGER)";
+    char *errmsg = NULL;
+    if (sqlite3_exec(db, sql, NULL, NULL, &errmsg) != SQLITE_OK)
+    {
+        sqlite3_free(errmsg);
+        return SQLITE_ERROR;
+    }
+    return SQLITE_OK;
+}
+
+/**
+ * close_database - Closes the SQLite database.
+ */
+void close_database()
+{
+    if (db)
+    {
+        sqlite3_close(db);
+        db = NULL;
+    }
+}
+
+/**
+ * free_account - Frees all memory associated with an Account struct.
+ */
+void free_account(Account *acc)
+{
+    if (!acc)
+    {
+        return;
+    }
+    free(acc->name);
+    free(acc->email);
+    free(acc->password);
+    if (acc->channels)
+    {
+        for (char **c = acc->channels; *c; c++)
+        {
+            free(*c);
+        }
+        free(acc->channels);
+    }
+    free_metadata(acc->metadata_head);
+    free(acc);
+}
+
+/**
+ * write_account_to_db - Appends an Account to the account database.
+ * Returns 1 on success, 0 on failure.
+ */
+int write_account_to_db(const Account *acc)
+{
+    if (find_account(acc->name))
+    {
+        // Account already exists
+        return 0;
+    }
+    const char *sql = "INSERT INTO accounts (name, email, password, time_registered, verified) VALUES (?, ?, ?, ?, ?)";
+    sqlite3_stmt *stmt;
+
+    if (sqlite3_prepare_v2(db, sql, -1, &stmt, NULL) != SQLITE_OK)
+    {
+        return 0;
+    }
+
+    sqlite3_bind_text(stmt, 1, acc->name, -1, SQLITE_STATIC);
+    sqlite3_bind_text(stmt, 2, acc->email, -1, SQLITE_STATIC);
+    sqlite3_bind_text(stmt, 3, acc->password, -1, SQLITE_STATIC);
+    sqlite3_bind_int(stmt, 4, (int)acc->time_registered);
+    sqlite3_bind_int(stmt, 5, acc->verified);
+
+    int result = sqlite3_step(stmt);
+    sqlite3_finalize(stmt);
+    return result == SQLITE_DONE ? 1 : 0;
+}
+
+/**
+ * read_accounts_from_db - Reads all accounts from the database.
+ * Returns a NULL-terminated array of Account*.
+ */
+Account **read_accounts_from_db(const char *name)
+{
+    const char *sql_all = "SELECT * FROM accounts";
```

Comment:
Do we really need this full table scan? Oo

Reply:
this is only a full table scan if you put NULL as the parameter for name

Reply:
Sorry, what I mean is do we use this with NULL anywhere?

Reply:
Yes, last night I wrote the RPC function for listing accounts, and also "LISTACC" debugging command uses that